### ISSUE TITLE
Fix OBB training decode gradients near FP32 subnormal dimensions

### ISSUE THREAD

**Issue by shapovalovei** [shapovalovei/YOLOX-OBB#38]
## Phase

Implementation authorized.

## Source validation

Issue #37 proved a generic decode-gradient arithmetic defect in the maintained OBB training path after the KLD numerical fix from #33 / PR #36.

## Proven defect

The maintained training decode uses:

`decoded_dimension = exp(raw_dimension) * stride`

For finite negative FP32 raw logits near the subnormal region, decoded width/height remains finite and positive and merged #36 KLD produces a finite decoded-dimension gradient, but PyTorch backward through the decode graph can produce `Inf` raw gradients before `exp(raw)` underflows to zero.

Representative stride-8 width case from #37:

- raw `-97`;
- decoded width `5.975136652e-42` (positive FP32 subnormal);
- KLD `dL/d(decoded_width) = -5.391932560e37` (finite);
- maintained raw gradient = `-inf`;
- FP64 mathematical raw gradient ≈ `-3.221771260e-4` (finite).

The scalar decomposition proved the causal arithmetic:

- upstream decoded gradient × stride overflows FP32;
- upstream decoded gradient × decoded dimension is finite;
- the maintained graph is `MulBackward0 -> ExpBackward0`;
- a validation-only algebraically equivalent training decode `exp(raw + log(stride))` preserves a finite raw gradient at the proven failure points.

The defect was reproduced for width/height and strides 8/16/32. No positive-side backward defect was found before forward decode overflow.

## Goal

Fix only the differentiable OBB **training decode** so finite mathematical raw-logit gradients are not lost due to intermediate backward overflow, while preserving the maintained model/output contracts and keeping inference/export behavior unchanged.

## Scope

Primary scope:

- `YOLOXHeadOBB_KLD.get_output_and_grid` width/height training decode;
- focused training decode-gradient regressions;
- ordinary forward-parity controls;
- KLD/assignment/target-validity regressions as compatibility checks.

A minimal local helper is allowed if it keeps intent clear and prevents duplication.

## Required implementation approach

Use:

`failing raw->decode->KLD->backward regression -> smallest algebraically equivalent training-only decode fix -> passing boundary regressions -> ordinary forward parity -> broader maintained tests`

Use #37's validated evidence as the starting hypothesis, not as permission to broaden scope.

## Acceptance criteria

1. Reproduce the current FP32 decode-gradient failure before the production change for width and height at representative subnormal raw logits.
2. Cover maintained strides 8, 16, and 32.
3. After the fix, raw gradients are finite at the #37 failure points and closely match the high-precision mathematical reference.
4. No artificial clamp/floor/`nan_to_num`/detach/straight-through workaround.
5. Ordinary training decoded values remain equivalent within a justified tight FP32 tolerance; quantify maximum absolute/relative difference over a deterministic ordinary matrix.
6. Document changed extreme representability boundaries if the selected equivalent formulation changes them.
7. Preserve exact-zero/underflow limitations honestly; do not claim recovery after `exp` information is irreversibly lost.
8. Preserve `decode_outputs` eager inference behavior unchanged.
9. Preserve external NumPy raw decode behavior unchanged.
10. Preserve raw ONNX/export output contract unchanged; no export implementation changes.
11. Preserve #36 KLD implementation unchanged.
12. Preserve #34 target-validity policy unchanged.
13. No head parameterization/objective/architecture redesign.
14. Run focused and broader local tests and report environment-only failures honestly.

## Design boundary

The accepted implementation boundary is **training-only**. The validation candidate `exp(raw + log(stride))` is mathematically equivalent and removed the proven gradient overflow, but it changes extreme FP32 representability boundaries and introduces small ordinary rounding differences. Therefore:

- do not apply the change to eager inference without separate evidence/authorization;
- do not alter external raw decode;
- do not alter ONNX raw-output behavior;
- quantify training-forward parity and boundary movement explicitly.

If the selected fix would require changing inference/export/public decode semantics, STOP and return to the orchestrator.

## Non-goals

- KLD changes;
- target-policy changes;
- training/retraining or mAP evaluation;
- model-quality claims;
- inference decode rewrite;
- external NumPy decode rewrite;
- ONNX/export implementation changes;
- head parameterization changes;
- optimizer/gradient clipping;
- assignment-memory work;
- MixUp;
- CI work;
- SDK changes.

## GitHub workflow

- start from current `origin/main` after approved preceding PRs are merged;
- create a dedicated branch;
- regression first;
- implement only this Issue scope;
- commit/push;
- open a Draft PR using `Refs` to this Issue and source validation #37;
- post factual implementation evidence;
- do not merge or close the Issue yourself; owner/orchestrator controls acceptance.

**Comment by shapovalovei** [comment]
## Implementation handoff

### Base

- Repository: `shapovalovei/YOLOX-OBB`
- Base SHA: `69c2c02cd5a7d8ae84bbce47834a22b973db12e5` (`origin/main` after
  `git fetch origin`); the required #33 / PR #36 merge is present.
- Branch: `fix/38-training-decode-gradient`
- Implementation commit: `ff030c0` (`fix: stabilize OBB training decode
  gradients (Refs #38)`)
- Environment: `/Users/eugene/creditCardReader/venv/bin/python`, Python
  3.11.11, PyTorch 2.2.2, CPU. CUDA was unavailable.
- Worktree was clean before branching. `AGENTS.md` remains untracked and
  excluded by `.git/info/exclude`.

### Proven old defect

Issue #37's representative stride-8 width case used the maintained target and
reported raw `-97`, decoded width `5.975136652e-42`, finite KLD
`dL/d(decoded_width) = -5.391932560e37`, old raw gradient `-inf`, and FP64
mathematical raw gradient approximately `-3.221771260e-4`. The corresponding
height case also had finite KLD decoded-gradient and old raw gradient `-inf`.

The regression was then run before changing production code on this base. It
exercised `raw tensor requiring grad -> get_output_and_grid -> merged KLD ->
backward` at raw `-97` for width and height at strides 8, 16, and 32. All six
subcases failed the finite-raw-gradient assertion on the old implementation.

For an independently reproducible centered target `[0, 0, 40, 10, 0]`, the
before/after/reference values were:

| stride | dimension | old decoded | old KLD dL/ddecoded | old raw grad | new raw grad | FP64 raw grad |
|---:|:---:|---:|---:|---:|---:|---:|
| 8 | width | `5.975136652e-42` | `-5.388747714e37` | `-inf` | `-3.219870850e-4` | `-3.219868557e-4` |
| 8 | height | `5.975136652e-42` | `-5.410960502e37` | `-inf` | `-3.233143070e-4` | `-3.233140407e-4` |
| 16 | width | `1.195027330e-41` | `-2.710179178e37` | `-inf` | `-3.238759818e-4` | `-3.238756623e-4` |
| 16 | height | `1.195027330e-41` | `-2.755564366e37` | `-inf` | `-3.292996844e-4` | `-3.292993916e-4` |
| 32 | width | `2.390054661e-41` | `-1.310694817e37` | `-inf` | `-3.132649581e-4` | `-3.132649540e-4` |
| 32 | height | `2.390054661e-41` | `-1.400095749e37` | `-inf` | `-3.346324374e-4` | `-3.346324650e-4` |

The old decoded dimensions and old KLD derivatives are finite; only the
maintained decode-chain raw gradient is non-finite. The largest absolute
new-vs-FP64 gradient difference in this six-case matrix is approximately
`3.20e-7`.

### Root cause

The old graph was:

```text
raw -> exp(raw) -> multiply by stride -> KLD -> backward
```

At the affected subnormal points, backward materialized the finite decoded
gradient times stride before multiplying by `exp(raw)`. That intermediate
overflowed FP32 even though the final mathematical product
`dL/d(decoded) * decoded` was finite. This is the backward operation-ordering
defect identified by #37, not a KLD defect.

### Implementation

Only `YOLOXHeadOBB_KLD.get_output_and_grid` changed its vectorized width/height
decode from:

```python
torch.exp(output[..., 2:4]) * stride
```

to:

```python
torch.exp(output[..., 2:4] + math.log(float(stride)))
```

The scalar `math.log` is evaluated from the maintained scalar stride. Adding a
Python scalar keeps the tensor dtype/device and creates no CPU tensor or
per-element Python loop. Folding the scalar into the exponent changes the
backward arithmetic ordering only in this training path.

### Gradient correction

At raw `-97`, the new gradients in the matrix above are finite and close to
the independent FP64 prediction-first KLD reference. The permanent focused
regressions also cover raw `-96.5`, `-96.75`, and `-97` for width/height at
strides 8/16/32 and require positive finite decoded dimensions, finite KLD,
and finite non-zero raw gradients.

### Stride/dimension matrix

All six required combinations pass after the fix:

- stride 8: width and height;
- stride 16: width and height;
- stride 32: width and height.

The FP64 reference uses the same raw, stride, centered target, prediction-first
KLD mathematics, degree-angle convention, and `taf=1` as the FP32 chain. It is
an independent direct FP64 mathematical expression, not a synthetic
multiplication-only test.

### Ordinary forward parity

A deterministic 161-sample raw matrix over `[-4, 4]` for each stride and both
width/height channels compared the new training decode with `exp(raw) *
stride`:

- maximum absolute difference: `3.662109375e-4` (stride 32, raw
  `3.799999952`, width);
- maximum relative difference: `3.551854775e-7` (stride 32, raw
  `0.700000048`, width).

These remain in the tight FP32-equivalence regime established by #37.

### Ordinary gradient parity

The deterministic ordinary gradient matrix used 15 raw values spanning
`[-4, 4]`, strides 8/16/32, and width/height, with the same merged KLD path
compared against the old mathematical `exp(raw) * stride` chain.

- maximum absolute raw-gradient difference: `2.533197403e-7` (raw `1.1`,
  stride 16, width; new `0.1952061653`, old `0.1952064186`).

### Representability boundaries

The training-only algebraic form necessarily changes extreme FP32
representability and this is explicitly tested:

| stride | old first exact-zero raw | new first exact-zero raw | old first +Inf raw | new first +Inf raw |
|---:|---:|---:|---:|---:|
| 8 | `-103.972084045` | `-106.051528931` | `86.643402100` | `86.643394470` |
| 16 | `-103.972084045` | `-106.744674683` | `85.950256348` | `85.950248718` |
| 32 | `-103.972084045` | `-107.437820435` | `85.257110596` | `85.257102966` |

The new exact-zero limit is not claimed to recover a gradient: the focused
test uses raw `-108`, where the new decode is itself exactly zero and the
selected raw gradient is zero. No clamp or floor was added. The positive-side
overflow boundary moves by one FP32 raw step; no new backward failure was
observed before forward decode overflow.

### Contracts preserved

- Eager `decode_outputs` remains `torch.exp(raw) * stride`; its focused
  forward-contract guard passes.
- External NumPy raw decoding remains unchanged.
- `tools/export_onnx.py` remains unchanged and continues to disable in-graph
  decode; raw ONNX output fields remain unchanged.
- `yolox/models/KLD_loss.py` remains unchanged; merged #36 numerical tests
  pass.
- #34 target-validity/preprocessing policy remains unchanged; its regressions
  pass.
- The generic HBB decode has the analogous expression but was intentionally
  not changed under #38.

### Tests

- Regression-first before production change: 1 focused test with 6 subcases,
  all six old raw gradients non-finite as expected for the reproduced defect.
- Focused after fix: `python -m unittest discover -s tests -p
  'test_obb_training_decode_gradients.py' -v` — 9 passed.
- `python -m py_compile yolox/models/yolo_head_obb_kld.py
  tests/test_obb_training_decode_gradients.py` — passed.
- `test_kld_numerical_stability.py` — 18 passed.
- `test_kld_prediction_target_contract.py` — 5 passed.
- `test_obb_assignment_geometry.py` — 5 passed.
- `test_obb_target_validity.py` — 11 passed.
- `test_obb_augmentation_geometry.py` — 9 passed.
- `test_obb_mixup_contract.py` — 1 passed.
- `test_obb_postprocess_contract.py` — 3 passed.
- `test_obb_dynamic_export.py` — 7 passed and 6 skipped by the existing
  ONNX/model/native dependency gate (`_polyiou` unavailable).
- Full local discovery: 87 tests; 79 passed, 6 skipped, and 2 pre-existing
  import errors from unavailable `DOTA_devkit_YOLO._polyiou` in
  `test_numpy2_rotated_postprocess.py` and `test_polyiou_packaging.py`.
- `git diff --check` — passed.

### Scope

Tracked changes are limited to the training decode expression and its focused
regression. No KLD, inference decode, external decoder, exporter, raw-output
contract, HBB, parameterization, target policy, assignment, training,
retraining, mAP/model-quality evaluation, optimizer/clipping, MixUp, CI, or
SDK work was performed.

### FACT / INFERENCE / UNVERIFIED

- FACT: the old real training chain produces non-finite raw gradients at the
  required FP32 subnormal cases while decoded values, KLD, and the independent
  mathematical gradient remain finite.
- FACT: the training-only log-space stride formulation passes the permanent
  six-case gradient, FP64 reference, ordinary parity, exact-zero, boundary,
  and eager-contract tests.
- FACT: only `yolox/models/yolo_head_obb_kld.py` and
  `tests/test_obb_training_decode_gradients.py` changed; KLD/export/external
  decoder sources are unchanged.
- INFERENCE: the causal defect is backward intermediate overflow caused by
  the old multiply-after-exp graph ordering, as established by #37's scalar
  decomposition and reproduced by the old-chain regression.
- UNVERIFIED: CUDA, CUDA AMP/Apex, MPS, decoded-output export graphs, real
  training incidence, convergence, and model-quality impact. No training or
  model-quality evaluation was run.

### PR

Draft PR: https://github.com/shapovalovei/YOLOX-OBB/pull/39

No merge or Issue closure performed.

**Comment by shapovalovei** [comment]
Orchestrator acceptance: PR #39 was independently reviewed and squash-merged at `c8d8b5611cab1b2f706a1638f53891b77b93dce0`. The implementation is accepted as a training-only arithmetic fix: `get_output_and_grid` folds the scalar stride into the exponent, the focused raw→decode→KLD→backward regressions pass for strides 8/16/32 and width/height, ordinary forward/gradient parity remains tight, and eager inference / external NumPy decode / raw ONNX export / KLD / target-validity contracts remain unchanged. No training or model-quality claim is made.
